File: app/src/main/java/pl/mosenko/sunnypodlaskie/persistence/dao/RxDao.java
package pl.mosenko.sunnypodlaskie.persistence.dao;

import com.j256.ormlite.dao.CloseableIterator;
import com.j256.ormlite.dao.CloseableWrappedIterable;
import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.dao.ForeignCollection;
import com.j256.ormlite.dao.GenericRawResults;
import com.j256.ormlite.dao.ObjectCache;
import com.j256.ormlite.dao.RawRowMapper;
import com.j256.ormlite.dao.RawRowObjectMapper;
import com.j256.ormlite.field.DataType;
import com.j256.ormlite.field.FieldType;
import com.j256.ormlite.stmt.GenericRowMapper;
import com.j256.ormlite.stmt.PreparedDelete;
import com.j256.ormlite.stmt.PreparedQuery;
import com.j256.ormlite.stmt.PreparedUpdate;
import com.j256.ormlite.support.ConnectionSource;
import com.j256.ormlite.support.DatabaseConnection;
import com.j256.ormlite.support.DatabaseResults;

import java.sql.SQLException;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;

import io.reactivex.Observable;

/**
 * Created by syk on 17.05.17.
 */

public interface RxDao<T, ID> extends Dao<T, ID> {

    Observable<T> rxQueryForId(final ID id);

    Observable<T> rxQueryForFirst(final PreparedQuery<T> preparedQuery);

    Observable<List<T>> rxQueryForAll();

    Observable<List<T>> rxQueryForEq(final String fieldName, final Object value);

    Observable<List<T>> rxQueryForMatching(final T matchObj);

    Observable<List<T>> rxQueryForMatchingArgs(final T matchObj);

    Observable<List<T>> rxQueryForFieldValues(final Map<String, Object> fieldValues);

    Observable<List<T>> rxQueryForFieldValuesArgs(final Map<String, Object> fieldValues);

    Observable<T> rxQueryForSameId(final T data);

    Observable<List<T>> rxQuery(final PreparedQuery<T> preparedQuery);

    Observable<Integer> rxCreate(final T data);

    Observable<T> rxCreateIfNotExists(final T data);

    Observable<CreateOrUpdateStatus> rxCreateOrUpdate(final T data);

    Observable<Integer> rxUpdate(final T data);

    Observable<Integer> rxUpdateId(final T data, final ID newId);

    Observable<Integer> rxUpdate(final PreparedUpdate<T> preparedUpdate);

    Observable<Integer> rxRefresh(final T data);

    Observable<Integer> rxDelete(final T data);

    Observable<Integer> rxDeleteById(final ID id);

    Observable<Integer> rxDelete(final Collection<T> datas);

    Observable<Integer> rxDeleteIds(final Collection<ID> ids);

    Observable<Integer> rxDelete(final PreparedDelete<T> preparedDelete);

    Observable<CloseableIterator<T>> rxIterator();

    Observable<CloseableIterator<T>> rxIterator(final int resultFlags);

    Observable<CloseableIterator<T>> rxIterator(final PreparedQuery<T> preparedQuery);

    Observable<CloseableIterator<T>> rxIterator(final PreparedQuery<T> preparedQuery, final int resultFlags);

    Observable<CloseableWrappedIterable<T>> rxGetWrappedIterable();

    Observable<CloseableWrappedIterable<T>> rxGetWrappedIterable(final PreparedQuery<T> preparedQuery);

    Observable<GenericRawResults<String[]>> rxQueryRaw(final String query, final String... arguments);

    <UO> Observable<GenericRawResults<UO>> rxQueryRaw(final String query, final RawRowMapper<UO> mapper, final String... arguments);

    <UO> Observable<GenericRawResults<UO>> rxQueryRaw(final String query, final DataType[] columnTypes, final RawRowObjectMapper<UO> mapper,
                                                      final String... arguments);

    Observable<GenericRawResults<Object[]>> rxQueryRaw(final String query, final DataType[] columnTypes, final String... arguments);

    Observable<Long> rxQueryRawValue(final String query, final String... arguments);

    Observable<Integer> rxExecuteRaw(final String statement, final String... arguments);

    Observable<Integer> rxExecuteRawNoArgs(final String statement);

    Observable<Integer> rxUpdateRaw(final String statement, final String... arguments);

    <CT> Observable<CT> rxCallBatchTasks(final Callable<CT> callable) throws Exception;

    Observable<Boolean> rxObjectsEqual(final T data1, final T data2);

    Observable<ID> rxExtractId(final T data);

    Observable<Class<T>> rxGetDataClass();

    Observable<FieldType> rxFindForeignFieldType(final Class<?> clazz);

    Observable<Boolean> rxIsUpdatable();

    Observable<Boolean> rxIsTableExists();

    Observable<Long> rxCountOf();

    Observable<Long> rxCountOf(final PreparedQuery<T> preparedQuery);

    <FT> Observable<ForeignCollection<FT>> rxGetEmptyForeignCollection(final String fieldName);

    Observable<ObjectCache> rxGetObjectCache();

    Observable<T> rxMapSelectStarRow(final DatabaseResults results);

    Observable<GenericRowMapper<T>> rxGetSelectStarRowMapper();

    Observable<RawRowMapper<T>> rxGetRawRowMapper();

    Observable<Boolean> rxIdExists(final ID id);

    Observable<DatabaseConnection> rxStartThreadConnection();

    Observable<Boolean> rxIsAutoCommit(final DatabaseConnection connection);

    Observable<ConnectionSource> rxGetConnectionSource();
}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/persistence/dao/RxDao.kt
package pl.mosenko.sunnypodlaskie.persistence.dao

import com.j256.ormlite.dao.*
import com.j256.ormlite.dao.Dao.CreateOrUpdateStatus
import com.j256.ormlite.field.DataType
import com.j256.ormlite.field.FieldType
import com.j256.ormlite.stmt.GenericRowMapper
import com.j256.ormlite.stmt.PreparedDelete
import com.j256.ormlite.stmt.PreparedQuery
import com.j256.ormlite.stmt.PreparedUpdate
import com.j256.ormlite.support.ConnectionSource
import com.j256.ormlite.support.DatabaseConnection
import com.j256.ormlite.support.DatabaseResults
import io.reactivex.Observable
import java.util.concurrent.Callable

/**
 * Created by syk on 17.05.17.
 */
interface RxDao<T, ID> : Dao<T, ID> {
    fun rxQueryForId(id: ID?): Observable<T?>?
    fun rxQueryForFirst(preparedQuery: PreparedQuery<T?>?): Observable<T?>?
    fun rxQueryForAll(): Observable<MutableList<T>>
    fun rxQueryForEq(fieldName: String?, value: Any?): Observable<MutableList<T?>?>?
    fun rxQueryForMatching(matchObj: T?): Observable<MutableList<T?>?>?
    fun rxQueryForMatchingArgs(matchObj: T?): Observable<MutableList<T?>?>?
    fun rxQueryForFieldValues(fieldValues: MutableMap<String?, Any?>?): Observable<MutableList<T?>?>?
    fun rxQueryForFieldValuesArgs(fieldValues: MutableMap<String?, Any?>?): Observable<MutableList<T?>?>?
    fun rxQueryForSameId(data: T?): Observable<T?>?
    fun rxQuery(preparedQuery: PreparedQuery<T?>?): Observable<MutableList<T?>?>?
    fun rxCreate(data: T?): Observable<Int?>?
    fun rxCreateIfNotExists(data: T?): Observable<T?>?
    fun rxCreateOrUpdate(data: T?): Observable<CreateOrUpdateStatus?>?
    fun rxUpdate(data: T?): Observable<Int?>?
    fun rxUpdateId(data: T?, newId: ID?): Observable<Int?>?
    fun rxUpdate(preparedUpdate: PreparedUpdate<T?>?): Observable<Int?>?
    fun rxRefresh(data: T?): Observable<Int?>?
    fun rxDelete(data: T?): Observable<Int?>?
    fun rxDeleteById(id: ID?): Observable<Int?>?
    fun rxDelete(datas: MutableCollection<T?>?): Observable<Int?>?
    fun rxDeleteIds(ids: MutableCollection<ID?>?): Observable<Int?>?
    fun rxDelete(preparedDelete: PreparedDelete<T?>?): Observable<Int?>?
    fun rxIterator(): Observable<CloseableIterator<T?>?>?
    fun rxIterator(resultFlags: Int): Observable<CloseableIterator<T?>?>?
    fun rxIterator(preparedQuery: PreparedQuery<T?>?): Observable<CloseableIterator<T?>?>?
    fun rxIterator(preparedQuery: PreparedQuery<T?>?, resultFlags: Int): Observable<CloseableIterator<T?>?>?
    fun rxGetWrappedIterable(): Observable<CloseableWrappedIterable<T?>?>?
    fun rxGetWrappedIterable(preparedQuery: PreparedQuery<T?>?): Observable<CloseableWrappedIterable<T?>?>?
    fun rxQueryRaw(query: String?, vararg arguments: String?): Observable<GenericRawResults<Array<String?>?>?>?
    fun <UO> rxQueryRaw(query: String?, mapper: RawRowMapper<UO?>?, vararg arguments: String?): Observable<GenericRawResults<UO?>?>?
    fun <UO> rxQueryRaw(query: String?, columnTypes: Array<DataType?>?, mapper: RawRowObjectMapper<UO?>?,
                        vararg arguments: String?): Observable<GenericRawResults<UO?>?>?

    fun rxQueryRaw(query: String?, columnTypes: Array<DataType?>?, vararg arguments: String?): Observable<GenericRawResults<Array<Any?>?>?>?
    fun rxQueryRawValue(query: String?, vararg arguments: String?): Observable<Long?>?
    fun rxExecuteRaw(statement: String?, vararg arguments: String?): Observable<Int?>?
    fun rxExecuteRawNoArgs(statement: String?): Observable<Int?>?
    fun rxUpdateRaw(statement: String?, vararg arguments: String?): Observable<Int?>?

    @Throws(Exception::class)
    fun <CT> rxCallBatchTasks(callable: Callable<CT?>?): Observable<CT?>?
    fun rxObjectsEqual(data1: T?, data2: T?): Observable<Boolean?>?
    fun rxExtractId(data: T?): Observable<ID?>?
    fun rxGetDataClass(): Observable<Class<T?>?>?
    fun rxFindForeignFieldType(clazz: Class<*>?): Observable<FieldType?>?
    fun rxIsUpdatable(): Observable<Boolean?>?
    fun rxIsTableExists(): Observable<Boolean?>?
    fun rxCountOf(): Observable<Long?>?
    fun rxCountOf(preparedQuery: PreparedQuery<T?>?): Observable<Long?>?
    fun <FT> rxGetEmptyForeignCollection(fieldName: String?): Observable<ForeignCollection<FT?>?>?
    fun rxGetObjectCache(): Observable<ObjectCache?>?
    fun rxMapSelectStarRow(results: DatabaseResults?): Observable<T?>?
    fun rxGetSelectStarRowMapper(): Observable<GenericRowMapper<T?>?>?
    fun rxGetRawRowMapper(): Observable<RawRowMapper<T?>?>?
    fun rxIdExists(id: ID?): Observable<Boolean?>?
    fun rxStartThreadConnection(): Observable<DatabaseConnection?>?
    fun rxIsAutoCommit(connection: DatabaseConnection?): Observable<Boolean?>?
    fun rxGetConnectionSource(): Observable<ConnectionSource?>?
}
